diff -urN percy++-1.0.0/CMakeLists.txt percy++-1.0.0-dp5/CMakeLists.txt
--- percy++-1.0.0/CMakeLists.txt	1970-01-01 01:00:00.000000000 +0100
+++ percy++-1.0.0-dp5/CMakeLists.txt	2014-10-16 16:16:18.826925493 +0100
@@ -0,0 +1,16 @@
+project (Percy++)
+
+find_file(NTL_INCLUDE_DIR "NTL/")
+find_path(NTL_INCLUDE_PARENTDIR "NTL/")
+
+include_directories(${NTL_INCLUDE_DIR} ${NTL_INCLUDE_PARENTDIR})
+add_library(percyclient
+	percyclient.cc itclient.cc percyparams.cc itparams.cc recover.cc
+	rsdecoder.cc percyio.cc FXY.cc gf2e.cc subset_iter.cc subset.cc portfolio.cc
+	agparams.cc agclient.cc percystats.cc streams.cc recursiveparams.cc
+	recursiveclient.cc hybridparams.cc nullclient.cc)
+add_library(percyserver
+	percyserver.cc itserver.cc percyparams.cc itparams.cc datastore.cc
+	percyio.cc gf2e.cc agparams.cc agserver.cc percystats.cc
+	streams.cc recursiveparams.cc recursiveserver.cc
+	hybridparams.cc)
